src/commands/validate.js

```diff
@@ -40,64 +43,52 @@ function getOverallErrors(variations = [], components = [], log, warn, error) {
 
-  if (!all) {
-    // this must come after config/pkgConf, so it can be supplied that way.
-    yargs.option('variations', {
-      demandOption: true,
-      requiresArg: true,
-    });
-    yargs.option('components', {
-      demandOption: true,
-      requiresArg: true,
-    });
-  }
-  yargs.positional('variations', {});
-
-  if (all && projectCount <= 0) {
-    throw chalk.red(`\`--all\` requires a non-empty “projects” config`);
+  if (all && allProjectNames.length <= 0) {
+    throw chalk.red('`--all` requires a non-empty “projects” config');
   }
   if (all && project) {
-    throw chalk.red(`\`--all\` and \`--project\` are mutually exclusive`);
+    throw chalk.red('`--all` and `--project` are mutually exclusive');
   }
   if (project && !has(projects, project)) {
     throw chalk.red(`Project "${project}" missing from “projects” config`);
   }
-  if (project) {
-    yargs.config(projects[project]);
+
+  if (projects) {
+    validateProjects(projects, allProjectNames, 'in the “projects” config');
+  } else {
+    validateProject(config);
   }
 };
-exports.handler = ({ variations, components, all, projects }) => {
-  if (all) {
-    const exitCodes = Object.entries(projects).map(([
-      project,
-      { variations, components, require: requires },
-    ]) => {
-      if (requires) { requireFiles(requires); }
-      try {
-        return getOverallErrors(
-          globToFiles(variations),
-          globToFiles(components),
-          x => console.log(`${chalk.inverse(chalk.blue(`Project “${project}”`))}: ${x}`),
-          x => console.warn(`${chalk.inverse(chalk.yellow(`Project “${project}”`))}: ${x}`),
-          x => console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${x}`),
-        );
-      } catch (e) {
-        console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${e.message}`);
-        return 1;
-      }
-    });
-    process.exit(Math.max(...exitCodes));
-  }
 
-  const exitCode = getOverallErrors(
+exports.handler = (config) => {
+  const { projects, projectNames } = normalizeConfig(config);
+
+  const exitCodes = [];
+
+  forEachProject(projects, projectNames, (project, {
     variations,
     components,
-    x => console.log(x),
-    x => console.warn(x),
-    x => console.error(x),
-  );
-  process.exit(exitCode);
+    require: requires,
+  }) => {
+    if (requires) { requireFiles(requires); }
+    try {
+      const exitCode = getOverallErrors(
+        globToFiles(variations),
+        globToFiles(components),
+        x => console.log(`${chalk.inverse(chalk.blue(`Project “${project}”`))}: ${x}`),
+        x => console.warn(`${chalk.inverse(chalk.yellow(`Project “${project}”`))}: ${x}`),
+        x => console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${x}`),
+      );
+      exitCodes.push(exitCode);
+    } catch (e) {
+      console.error(`${chalk.inverse(chalk.red(`Project “${project}”`))}: ${e.message}`);
+      exitCodes.push(1);
```

Comment:
It might be worth dropping a comment here explaining why you want to wait and run all of the things and collect the exit codes before exiting.